### DISCUSSION TITLE
Proper clean-up hook usable with pyproject-build

### DISCUSSION THREAD

**Discussion by ericzolf** [pypa/setuptools#4121] [Q&A] [answered]
After having read https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html and having already introduced a working pyproject.toml, I'm in the [process of replacing all invocations](https://github.com/rdiff-backup/rdiff-backup/tree/ericzolf-remove-setup) of `./setup.py` with `pyproject-build` aka `python -m build`.

So far so good, but I have custom build/clean steps in my `./setup.py` and I've noticed that at least the cleaning step(s) doesn't work, the generated files aren't removed after the sdist/wheel have been built, and pyproject-build doesn't have an explicit clean "target". The build step(s) seem to work correctly, but I also don't know how much is due to luck.

Any hint on how to fix my plug-ins would be appreciated.

**Comment by abravalheri** [comment]
Hi @ericzolf, is it possible for you to implement a `clean` task using a task runner like `tox`, `nox`, `pydoit`, `invoke`? (or any task runner really...)

One thing that I find very useful is to use [`git clean -fdx`](https://git-scm.com/docs/git-clean) (or a variation using `-X`, `-e`, etc...) to remove files that are not tracked by git.

**Reply by webknjaz** [reply]
I'd like to share that I'm making in-tree build backends for projects with C-extensions and I recently tried out an idea of building in a temporary directory. So I included some logic into my setuptools wrapper to copy files to a tmp directory for non-editable builds. This lets me be sure I haven't accidentally included any garbage into the wheel + doesn't populate generated files into my dev env in such cases: https://github.com/aio-libs/yarl/blob/8d28329/packaging/pep517_backend/_backend.py#L284-L293.

**Comment by ericzolf** [comment]
> Hi @ericzolf, is it possible for you to implement a `clean` task using a task runner like `tox`, `nox`, `pydoit`, `invoke`? (or any task runner really...)

That could be an approach, but I thought to keep it in the back-end. This remark though makes me think that I might be poking in the wrong place: the back-end (setuptools) already knows how to clean-up, it's possibly rather the front-end (build) which doesn't use it properly.
 
> One thing that I find very useful is to use [`git clean -fdx`](https://git-scm.com/docs/git-clean) (or a variation using `-X`, `-e`, etc...) to remove files that are not tracked by git.

Not an option in this case, I need something more specific, I can't risk lose all local changes each time I want to clean-up the build.

**Reply by abravalheri** [reply, marked as answer]
For now, the definition of PyPA for backend APIs does not include clean hooks. This might need to be discussed in the Python discourse.

**Reply by ericzolf** [reply]
Thanks, I'm then continuing the discussion at https://discuss.python.org/t/custom-build-targets-for-cleanup-post-pep-518/13741/32
